SPEC.md

```diff
@@ -0,0 +1,225 @@
+
+<p align="center">
+  <br>
+  <img src="images/apollo-fetcher.png" alt="Apollo Fetcher"/>
+</p>
+
+Fetchers are designed for current Apollo Client use-cases and to provide room for extensions. In addition to supporting Apollo Client, Apollo Fetchers are designed to be used independently as a very simple GraphQL client. These are the use cases that fetchers should support:
+
+* Do a simple GraphQL query with it that resolves only once
+* Do a simple GraphQL mutation that resolves only once
+* Do a subscription
+* Do a GraphQL query that contains @defer, @stream or @live
+* Send extra metadata along with queries
+* Support middleware to modify headers
+* Support middleware to modify query, variables, operationName and metadata
+* Isomorphic: It works in the browser and on the server
+* Support afterware to modify response
+* Support HTTP  transport, websocket transport, hybrid transport
+* Support alternative transports
+* Support per request retries based on network status or response
+* Support HTTP transport batching
+* Support polling
+* Support batching
+* Request deduplication of on-the-wire requests
+* Support splitting: eg. send queries and mutations to different servers
+
+Since this list focuses on Apollo Client's needs, please open a Pull Request adding to this list as you think of your desired use cases.
+
+In order to support all these use cases with a simple fetcher design, we want to use a modular architecture that allows fetchers to delegate to other fetchers.
+
+## Example
+
+This repository will contain a family of canonical fetchers that provide example extensions. The goal of these canonical fetchers is to be self-contained without changing the context of each operation.
+
+The current sample is http-fetcher, which sends an operation over http to a server at a specific URI. The uri and fetch function are passed to the constructor. This Http-fetcher can be used as a stand-alone GraphQL client:
+
+<p align="center">
+  <br>
+  <img src="images/http-fetcher.png" alt="Apollo Fetcher"/>
+</p>
+
+```js
+const httpFetcher = new HTTPFetcher({ uri: 'http://api.githunt.com/graphql' });
+
+const responseObservable = httpFetcher.request({
+  query,
+  variables,
+  operationName,
+  context,
+});
+
+responseObservable.subscribe({
+  next(data){ console.log('received data', data); },
+  error(error){ console.error(error); },
+  complete(){ console.error('request complete'); },
+});
+```
+
+A fetcher can delegate to other fetchers. For example a polling fetcher can send a query operation to an http fetcher on a specified interval. Each request is routed through that http fetcher and returned through the observable returned by the polling fetcher. Using this type of delegation, fetchers can form stacks that each add their own functionality.
+
+<p align="center">
+  <br>
+  <img src="images/polling-stack.png" alt="Http Fetcher"/>
+</p>
+
+
+```js
+const pollingFetcher = new PollingFetcher(
```

Comment:
I'd say we could go even further. 

```js
const network: Fetcher = HTTPFetcher({ uri: '...' });
const polling: Fetcher = PollingFetcher({ interval: 100 });

const fetcher = network.concat(polling);
// or
const fetcher = Fetcher.from([ network, polling ]);
// or
const fetcher = Fetcher.of(network).concat(polling);
```

If possible, does it make sense to stick to algebraic structures? Or is that too confusing for people? If you did, you can do things like this:

```js
// create the fetcher
const fetcher: Fetcher = HTTPFetcher({ uri:  'http://api.githunt.com/graphql' })
   // don't send requests with client directive
  .filter(request => !isClientOperation(request.operation))
  // add authorization (middleware)
  .map(request => {
    request.context.headers.authorization = userId();
    return request;
  });

// operationName can deduced from the operation
// this part won't normally be exposed to the end user but used in apollo-client or other libs
// should this be positional vs destructured
const observable = fetcher.request({ operation: DocumentNode, variables: Object, context: Object });

const subscription  = observable
  // don't notify if offline error
  .filter((data, error) => error instanceof NetworkError)
  // change shape of data (afterware)
  .map(data => data.items.map(x => localize(x.date)))
  .subscribe(
    // data
    data => console.log('received data', data),
    // error
    error => console.error(error),
    // complete
    () => console.error('request complete'),
  );

// When we're done
subscription.dispose();

```

From what I can tell from the (excellent!) diagrams, the critical bits are the request information and the result information. Each fetcher should be able to modify this information (forming middlewares and afterwares), and should be able to build on top of each other.

The `Fetcher` interface should allow for control flow based on the `Request` information (i.e operation, variables, context). One should be able to map, filter, and concat the Fetcher to modify request information before it is sent. Since the fetcher is returned from each function, a specific fetcher would be able to add things to the interface like `interval` or `batchIf` or etc. I think there is a lot that can be learned from RxJS in building enjoyable observable interfaces.

Upon `.request`

The `Observable` should allow for control flow related to the result of the request (i.e. data and error). One should be able to map, filter, and concat based on result and error data. This forms the basis for afterware and allows fine grained control of subscription handler firings.

In fact, I wonder if looking at something like [RxJS.lite](https://www.npmjs.com/package/rx-lite) might even be a good starting point for the generic Fetcher library. Things like `catch`, `interval`, and `retry` seem like perfect fits for fetchers.

I love the idea of fetchers being used on their own as a slim build of a graphql library. Apollo Client itself could be rebuilt to be essentially the ~fully built~ combination of excellent fetchers. Having more of a functional approach to the way Fetchers are defined and used could make this natural and allow for interop between compile to languages (like reason) and give strong type inference for things like typescript and flow.